asserts/ifacedecls.go

```diff
@@ -0,0 +1,187 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2015 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package asserts
+
+import (
+	"fmt"
+	"regexp"
+	"strconv"
+)
+
+type attrMatcher interface {
+	match(interface{}) error
+}
+
+// compileAttrMatcher compiles an attrMatcher derived from constraints,
+// context can be "top" or "map", alternatives have no influence on it
+func compileAttrMatcher(context string, constraints interface{}) (attrMatcher, error) {
+	switch x := constraints.(type) {
+	case map[string]interface{}:
+		return compileMapAttrMatcher(x)
+	case []interface{}:
+		// TODO: disallow nesting alternatives directly inside alernatives?
+		return compileAltAttrMatcher(context, x)
+	case string:
+		if context == "top" {
+			return nil, fmt.Errorf("first level of non alternative constraints must be a set of key-value contraints")
+		}
+		return compileRegexpAttrMatcher(x)
+	default:
+		return nil, fmt.Errorf("constraint must be a key-value map, regexp or a list of alternative constraints: %v", x)
+	}
+}
+
+type mapAttrMatcher map[string]attrMatcher
+
+func compileMapAttrMatcher(m map[string]interface{}) (attrMatcher, error) {
+	matcher := make(mapAttrMatcher)
+	for k, constraint := range m {
+		matcher1, err := compileAttrMatcher("map", constraint)
+		if err != nil {
+			return nil, fmt.Errorf("constraint for %q: %v", k, err)
+		}
+		matcher[k] = matcher1
+	}
+	return matcher, nil
+}
+
+func matchEntry(k string, matcher1 attrMatcher, v interface{}) error {
+	if v == nil {
+		return fmt.Errorf("%q has constraints but is unset", k)
```

Comment:
yes, I was thinking about this, if we need it soon it would be good to work out how it should look